src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+    };
+    ctx.call::<DocumentSymbolRequest, _>(
+        meta,
+        req_params,
+        move |ctx: &mut Context, meta, result| {
+            process_results_for_next_or_prev(meta, editor_params, result, ctx)
+        },
+    );
+}
+
+/// When this function is called we have got the results of DocumentSymbolRequest.
+/// This function does some minimal processing and then hands the data off to an
+/// appropriate function that will search for the next/previous symbol.
+///
+/// After that is done, we call `next_or_prev_response()` which does the job
+/// of actually sending the response back to Kakoune.
+///
+/// This function's code is derived from `document_symbol::editor_document_symbol()`.
+/// and its basic structure is similar to it.
+fn process_results_for_next_or_prev(
+    meta: EditorMeta,
+    editor_params: EditorParams,
+    result: Option<DocumentSymbolResponse>,
+    ctx: &mut Context,
+) {
+    let params = NextOrPrevSymbolParams::deserialize(editor_params).unwrap();
+    let hover = params.hover;
+    let maybe_details = match result {
+        Some(DocumentSymbolResponse::Flat(mut result)) => {
+            if result.is_empty() {
+                return;
+            }
+            // First let's sort the results so we can find next and previous properly
+            // This step does _not_ happen in `document_symbol::editor_document_symbol()`.
+            //
+            // Some language servers return symbol locations in unsorted order
+            // or non-ascending order.
+            result.sort_by(|a, b| a.location.range.start.cmp(&b.location.range.start));
+            get_next_or_prev_symbol_information_details(result, params, ctx)
+        }
+        Some(DocumentSymbolResponse::Nested(mut result)) => {
+            if result.is_empty() {
+                return;
+            }
+            // First let's sort the results so we can find next and previous properly
+            // This step does _not_ happen in `document_symbol::editor_document_symbol()`.
+            //
+            // Some language servers return symbol locations in unsorted order
+            // or non-ascending order.
+            result.sort_by(|a, b| a.range.start.cmp(&b.range.start));
+            get_next_or_prev_document_symbol_details(result, params, &meta, ctx)
+        }
+        None => {
+            return;
+        }
+    };
+
+    next_or_prev_response(meta, ctx, maybe_details, hover);
+}
+
+/// Send the response back to Kakoune. This could be either:
+/// a) Instructions to move the cursor to the next/previous symbol.
+/// b) Instructions to show hover information (without actually moving the
+///    visible user cursor) _of_ the next/previous symbol
+fn next_or_prev_response(
+    meta: EditorMeta,
+    ctx: &mut Context,
+    maybe_details: Option<(String, KakounePosition, String, SymbolKind)>,
+    hover: bool,
+) {
+    if let Some((filename, symbol_position, name, kind)) = maybe_details {
```

Comment:
I'd use an early return here, to reduce indentation and make it easier to read, especially for error handling.
I usually use this pattern

```rust
    let (filename, symbol_position, name, kind) = match maybe_details {
        Some((filename, symbol_position, name, kind)) => (filename, symbol_position, name, kind),
        None => {
            ctx.exec(...);
            return;
        }
    }
```

Same for the `if hover` below - IMO the easy case should come first, like `if !hover { ...; return; }`.